packages/project/src/merge/map-uuids.ts

```diff
@@ -39,82 +40,234 @@ export interface MappingResults {
-      else if (ex === undefined) edgeMapping[edgeId] = true;
+  for (const [parent, children] of Object.entries(sourceEdges)) {
+    for (const child of children) {
+      const tparent = idMap.has(parent) ? idMap.get(parent) : parent;
+      const tchild = idMap.has(child) ? idMap.get(child) : child;
+      const edgeKey = `${parent}-${child}`;
+      const targetEdgeKey = `${tparent}-${tchild}`;
+      if (edgeMapping[targetEdgeKey] === null)
+        edgeMapping[edgeKey] = getEdgeUuid(tparent, tchild, target.steps);
+      else edgeMapping[edgeKey] = true;
     }
   }
+
   return {
     nodes: nodeMapping,
     edges: edgeMapping,
   };
 };
 
-// if newchildren has some additional nodes, that's fine. it's the same children!
-// when a node is missing, nope, not the same children
-function sameChildren(mainChildren: string[], newChildren: string[]) {
-  if (newChildren.length < mainChildren) return false;
-  for (const child of mainChildren) {
-    if (!newChildren.includes(child)) return false;
+function getEdgeUuid(
+  parentId: string,
+  childId: string,
+  steps: Workflow['steps']
+) {
+  const parentNode = steps.find((step) => step.id === parentId);
+  if (!parentNode) return;
+  if (typeof parentNode.next !== 'object') return;
+  const edge = parentNode.next[childId];
+  return edge?.openfn?.uuid;
+}
+function getStepUuid(step: Workflow['steps'][number]) {
+  return step?.openfn?.uuid || step.id;
+}
+function getEdges(steps: Workflow['steps']) {
+  const edges: Record<string, string[]> = {};
+  for (const step of steps) {
+    const next =
+      typeof step.next === 'string' ? { [step.next]: true } : step.next || {};
+
+    for (const toNode of Object.keys(next)) {
+      if (!Array.isArray(edges[step.id])) edges[step.id] = [toNode];
+      else edges[step.id].push(toNode);
+    }
   }
-  return true;
+  return edges;
+}
+
+function getParent(id: string, steps: Workflow['steps']) {
+  const edges = getEdges(steps);
+  const found = Object.entries(edges).find(([parent, children]) =>
+    children.includes(id)
+  );
+  if (!found) return;
+  return found[0]; // getting the parent id
+}
+
+interface Pool {
+  source: Workflow['steps'];
+  target: Workflow['steps'];
 }
 
-function getNodesWithParent(parent: string, parentsMap: Map<string, string>) {
-  const matches: string[] = [];
-  for (const [child, child_parent] of parentsMap.entries()) {
-    if (parent === child_parent) matches.push(child as string);
+interface MapStepResult {
+  mapping: Record<string, string>;
+  idMap: Map<string, string>;
+  pool: Pool;
+}
+
+function mapStepsById(
+  source: Workflow['steps'],
+  target: Workflow['steps']
+): MapStepResult {
+  const targets: Record<string, Workflow['steps'][number]> = {};
+  const mapping: Record<string, string> = {};
+  const idMap = new Map<string, string>();
+
+  for (const target_step of target) {
+    targets[target_step.id] = target_step;
+    mapping[target_step.id] = null;
```

Comment:
I think I'm wrong about this bit - Lightning might have to worry about removal, but looking at it today I'm sure that we _don't_